### PR TITLE
 feat!: make the library usable without the builtin reverse proxy

### PR DESCRIPTION
**Pull request by dunglas**
* Make the library usable with any Go HTTP handler or reverse proxy, including Caddy server
* Improve the documentation
* Add an example
* Switch from Logrus to Uber Zap

### CODE REVIEW

File: vulcain.go [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+// Vulcain parses and allow to modify the HTTP response according to the "preload" and "fields" directives extracted from the request
+// Use New to create a Vulcain instance
+type Vulcain struct {
+	pushers *pushers
+	openAPI *openAPI
+	logger  *zap.Logger
+}
+
+// New creates a Vulcain instance
+func New(options Options) *Vulcain {
```
Comment (sagikazarmark) [inline vulcain.go:51]: Looks like this constructor could utilize [functional options](https://sagikazarmark.hu/blog/functional-options-on-steroids/) since most of the options are actually optional.

File: vulcain.go [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+	logger := options.Logger
+	if options.Logger == nil {
```
Comment (sagikazarmark) [inline vulcain.go:53]: Unless you have a special need for zap, I'd urge you to consider creating a custom logger interface and using it instead of hard-coding zap into your library. I maintain a couple large projects and I really hate the fact that I have to pull in 5 different loggers, because different libraries chose different logging solutions. Same thing as PSR-3 and PSR-18 for PHP. Kind of sad that each ecosystem needs to reinvent this wheel. :\

Comment (dunglas) [reply vulcain.go:53]: I'll consider this in a next PR, but I'm afraid that it will add some boilerplate code to the library, which I want to keep as small and as focused on its core business as possible. I used Zap for this POC because it's a requirement of Caddy (I also created a Caddy module relying on this PR that I will release soon).
The main pain point is that I need structured logging, and recreating an interface for this is a bit cumbersome. I hope that structured logging will land in the standard library at some point...

File: vulcain.go (hunk 1 above)
Comment (dunglas) [reply vulcain.go:51]: Nice! I'll do that.

File: vulcain.go (hunk 2 above)
Comment (sagikazarmark) [reply vulcain.go:53]: Personally, I'd make the Caddy module a separate Go module, so the library can be used without Caddy being a dependency.

There was an [attempt](https://docs.google.com/document/d/1shW9DZJXOeGbG9Mr9Us9MiaPqmlcVatD_D8lrOXRNMU) to make structured logging part of the standard library, but it failed because the authors wasn't confident enough that a generic interface for all use cases can be created that would make a standard library solution suitable for everyone. So I wouldn't wait for a stdlib solution.

There are two common approaches for structured logging:

- `map[string]interface{}` fields: [example](https://github.com/logur/logur/blob/c5ec959/logger.go#L9-L41)
- variadic `...interface{}` fields: [example](https://github.com/logur/logur/blob/c5ec959/logger.go#L171-L205)

Obviously, those are not as type-safe as Zap, but they provide a nice generic interface. From the library side, you don't actually need an implementation (other than a noop one for testing). Since Caddy already requires Zap, you can use an [adapter](https://github.com/logur/adapter-zap/) in the Caddy module.

To sum up:

- In the library all you need is an interface serving your needs (eg. you don't need trace level? don't add it to the interface) and a dummy implementation for testing. I don't think it's a lot of boilerplate.
- you need implementations where you actually use the library (ie. Caddy module). you can either use existing ones that implement the above interfaces (eg. https://github.com/logur?q=adapter-&type=&language= ) or just write a wrapper around the chosen library. (A wrapper for Zap is less than 100 LOC)

File: vulcain.go (hunk 1 above)
Comment (dunglas) [reply vulcain.go]: Done in de33341a2ca258cee1a2be11c98776824e1602a0. Very nice pattern!

Comment (sagikazarmark) [reply vulcain.go]: Yeah, I like it very much!

File: vulcain.go (hunk 2 above)
Comment (dunglas) [reply vulcain.go]: It already is a separate module (using this lib)! The only thing that leaked inside this lib is the use of Zap (but Logrus is deprecated anyway). I'm sure we'll find a way to get rid of the dependency to Zap at some point too.

Comment (dunglas) [reply vulcain.go]: Zap has some nice features we use that will be harder to cover with a simple interface: for instance we use the lazy call to String().

Comment (sagikazarmark) [reply vulcain.go]: > for instance we use the lazy call to String().

Can you elaborate what that means (I'm not that familiar with Zap)?

Comment (sagikazarmark) [reply vulcain.go]: > I'm sure we'll find a way to get rid of the dependency to Zap at some point too.

Feels like this is a good time, since you are changing the API anyway. 🙂

Comment (dunglas) [reply vulcain.go]: I mean we use this kind of things: https://github.com/uber-go/zap/blob/v1.16.0/field.go#L340
If we create a minimal interface, we'll have to extend it every time we use a new field type or a new level... and it will be a BC break requiring to bump the major version of the library. I don't think it is desirable too.

To be honest, I feel like assuming the dependency to Zap is not so bad because it's by far the most popular structured logger now that Logrus is in maintenance-mode.

File: openapi.go [hunk 3]
```diff
@@ -80,13 +88,15 @@ func (o *openAPI) generateLinkForResponse(response *openapi3.Response, selector,
...
+// generateLink uses the template IRI extracted from the OpenAPI description to generate a URL
 func (o *openAPI) generateLink(operationID, parameter, value string) string {
 	for path, i := range o.swagger.Paths {
 		if op := i.GetOperation("GET"); op != nil && op.OperationID == operationID {
 			return strings.ReplaceAll(path, "{"+parameter+"}", value)
 		}
 	}
 
-	log.WithField("operationId", operationID).Debug("Operation not found in the OpenAPI specification")
+	o.logger.Debug("peration not found in the OpenAPI specification", zap.String("operationID", operationID))
```
Comment (soyuka) [inline openapi.go:99]:
```suggestion
	o.logger.Debug("Operation not found in the OpenAPI specification", zap.String("operationID", operationID))
```

File: openapi.go [hunk 4]
```diff
@@ -23,18 +26,21 @@ func newOpenAPI(file string) *openAPI {
...
+		logger,
 	}
 }
 
+// getRoute gets the openapi3filter.Route instance related to the given URL
 func (o *openAPI) getRoute(url *url.URL) *openapi3filter.Route {
 	route, _, err := o.router.FindRoute("GET", url)
 	if err != nil {
-		log.WithFields(log.Fields{"url": url, "reason": err}).Debug("Route not found in the OpenAPI specification")
+		o.logger.Debug("route not found in the OpenAPI specification", zap.Stringer("url", url), zap.Error(err))
```
Comment (soyuka) [inline openapi.go:37]:
```suggestion
		o.logger.Debug("Route not found in the OpenAPI specification", zap.Stringer("url", url), zap.Error(err))
```

File: pusher.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+	if req.Header.Get(internalRequestHeader) != "" {
+		w.Done()
+		return
+	}
+
+	// Wait for subrequests to finish
+	w.Wait()
+	p.remove(w.id)
+}
```
Comment (soyuka) [inline pusher.go:152]: Nice to have this separated from the proxy!

File: openapi.go (hunk 4 above)
Comment (dunglas) [reply openapi.go:37]: it's better to use lower case everywhere to be consistent with what's done by Caddy.

**Review by soyuka — COMMENTED** [review]
I like the changes! I have the same concerns with Zap as lighten above. 

I proposed changes to `logger.Debug`, shouldn't you make all logs start with an uppercase letter for consistency?

File: vulcain.go (hunk 2 above)
Comment (sagikazarmark) [reply vulcain.go]: Well, how about just calling `String` manually on the value? :D

Comment (dunglas) [reply vulcain.go]: It will always be called then, even if we not enable the Debug mode :)

**Comment by dunglas** [comment]
To move forward I merged with Zap. I'll be glad to remove this dependency if we find a practical way to do it! PR welcome :)

File: vulcain.go (hunk 2 above)
Comment (sagikazarmark) [reply vulcain.go]: I guess this is something that the implementations should provide: if the value is a Stringer, call String on it when evaluated. Actually, logur adapters should already work this way. :)

**Comment by sagikazarmark** [comment]
TBH I'm more interested in Mercure, so I'd much rather work on a PR there, once the refactor is done.
